Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,687 @@
+    InitMKDD();
+  if (Common::swap16(udpdata->destination_port) == 53)
+  {
+    target = sf::IpAddress(m_dns_ip.c_str());  // dns server ip
+  }
+  else
+  {
+    target = sf::IpAddress(Common::swap32(*(u32*)ipdata->destination_addr));
+  }
+
+  network_ref[i].udp_socket.send(data, Common::swap16(udpdata->length) - 8, target,
+                                 Common::swap16(udpdata->destination_port));
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::SendFrame(const u8* frame, u32 size)
+{
+  Common::EthernetHeader* hwdata;
+  Common::IPv4Header* ipdata;
+  Common::UDPHeader* udpdata;
+  Common::TCPHeader* tcpdata;
+
+  int offset = 0;
+  std::lock_guard<std::mutex> lock(mtx);
+  memmove(&m_out_frame[0], frame, size);
+  INFO_LOG_FMT(SP1, "Sending packet {:x}\n{}", size, ArrayToString(&m_out_frame[0], size, 16));
+  // handle the packet data
+  hwdata = (Common::EthernetHeader*)&m_out_frame[0];
+  if (hwdata->ethertype == 0x08)  // IPV4
+  {
+    // IP sub
+    ipdata = (Common::IPv4Header*)&m_out_frame[14];
+    switch (ipdata->protocol)
+    {
+    case IPPROTO_UDP:
+      udpdata = (Common::UDPHeader*)&m_out_frame[14 + ((ipdata->version_ihl & 0xf) * 4)];
+      offset = 14 + ((ipdata->version_ihl & 0xf) * 4) + 8;
+      if (Common::swap16(udpdata->destination_port) == 67)
+      {
+        Common::DHCPBody* request = (Common::DHCPBody*)&m_out_frame[offset];
+        HandleDHCP(hwdata, udpdata, request);
+      }
+      else
+      {
+        HandleUDPFrame(hwdata, ipdata, udpdata, &m_out_frame[offset]);
+      }
+      break;
+
+    case IPPROTO_TCP:
+      tcpdata = (Common::TCPHeader*)&m_out_frame[14 + ((ipdata->version_ihl & 0xf) * 4)];
+      offset = 14 + ((ipdata->version_ihl & 0xf) * 4) + (((tcpdata->properties >> 4) & 0xf) * 4);
+      HandleTCPFrame(hwdata, ipdata, tcpdata, &m_out_frame[offset]);
+      break;
+    }
+  }
+  if (hwdata->ethertype == 0x608)  // arp
+  {
+    Common::ARPHeader* arpdata = (Common::ARPHeader*)&m_out_frame[14];
+    HandleARP(hwdata, arpdata);
+  }
+
+  m_eth_ref->SendComplete();
+  return true;
+}
+
+size_t TryGetDataFromSocket(StackRef* ref, u8* buffer)
+{
+  Common::EthernetHeader* hwdata;
+  Common::IPv4Header* ipdata;
+  Common::UDPHeader* udpdata;
+  Common::TCPHeader* tcpdata;
+  size_t datasize = 0;
+  unsigned short prt;
+  int l;
+
+  switch (ref->type)
+  {
+  case IPPROTO_UDP:
+    ref->udp_socket.receive(&buffer[0x2a], 1500, datasize, ref->target, prt);
+    if (datasize > 0)
+    {
+      memset(&buffer[0], 0, 0x2a);
+      hwdata = (Common::EthernetHeader*)&buffer[0];
+      ipdata = (Common::IPv4Header*)&buffer[14];
+      udpdata = (Common::UDPHeader*)&buffer[0x22];
+
+      // build header
+      *hwdata = Common::EthernetHeader(IP_PROTOCOL);
+      hwdata->destination = ref->bba_mac;
+      hwdata->source = ref->my_mac;
+
+      ref->from.sin_port = htons(prt);
+      ref->from.sin_addr.S_un.S_addr = htonl(ref->target.toInteger());
```

Comment:
I think the portable way of doing this is using `sin_addr.s_addr`. The Linux builds are complaining about this.

Reply:
and macOS builds too

Reply:
yup, fixed it but havent pushed yet. Thx for testing, its more compatible than i originaly expected!